Mathlib/Algebra/Module/Injective.lean

```diff
@@ -0,0 +1,467 @@
+    (Set.mem_image _ _ _).mpr ⟨a, ha, rfl⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of.le_max Module.Baer.ExtensionOf.le_max
+
+variable (i f) [Fact <| Function.Injective i]
+
+instance ExtensionOf.inhabited : Inhabited (ExtensionOf i f) where
+  default :=
+    { domain := LinearMap.range i
+      toFun :=
+        { toFun := fun x => f x.2.choose
+          map_add' := fun x y => by
+            have eq1 : _ + _ = (x + y).1 := congr_arg₂ (· + ·) x.2.choose_spec y.2.choose_spec
+            rw [← map_add, ← (x + y).2.choose_spec] at eq1
+            dsimp
+            rw [← Fact.out (p := Function.Injective i) eq1, map_add]
+          map_smul' := fun r x => by
+            have eq1 : r • _ = (r • x).1 := congr_arg ((· • ·) r) x.2.choose_spec
+            rw [← LinearMap.map_smul, ← (r • x).2.choose_spec] at eq1
+            dsimp
+            rw [← Fact.out (p := Function.Injective i) eq1, LinearMap.map_smul] }
+      le := le_refl _
+      is_extension := fun m => by
+        simp only [LinearPMap.mk_apply, LinearMap.coe_mk]
+        dsimp
+        apply congrArg
+        exact Fact.out (p := Function.Injective i)
+          (⟨i m, ⟨_, rfl⟩⟩ : LinearMap.range i).2.choose_spec.symm }
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of.inhabited Module.Baer.ExtensionOf.inhabited
+
+/-- Since every nonempty chain has a maximal element, by Zorn's lemma, there is a maximal
+`extension_of i f`. -/
+def extensionOfMax : ExtensionOf i f :=
+  (@zorn_nonempty_partialOrder (ExtensionOf i f) _ ⟨Inhabited.default⟩ fun _ hchain hnonempty =>
+      ⟨ExtensionOf.max hchain hnonempty, ExtensionOf.le_max hchain hnonempty⟩).choose
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max Module.Baer.extensionOfMax
+
+theorem extensionOfMax_is_max :
+    ∀ a : ExtensionOf i f, extensionOfMax i f ≤ a → a = extensionOfMax i f :=
+  (@zorn_nonempty_partialOrder (ExtensionOf i f) _ ⟨Inhabited.default⟩ fun _ hchain hnonempty =>
+      ⟨ExtensionOf.max hchain hnonempty, ExtensionOf.le_max hchain hnonempty⟩).choose_spec
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_is_max Module.Baer.extensionOfMax_is_max
+
+-- Porting note: helper function. Lean looks for an instance of `Sup (Type u)` when the
+-- right hand side is substituted in directly
+@[reducible]
+def supExtensionOfMaxSingleton (y : N) : Submodule R N :=
+  (extensionOfMax i f).domain ⊔ (Submodule.span R {y})
+
+variable {f}
+
+private theorem extensionOfMax_adjoin.aux1 {y : N} (x : supExtensionOfMaxSingleton i f y) :
+    ∃ (a : (extensionOfMax i f).domain) (b : R), x.1 = a.1 + b • y := by
+  have mem1 : x.1 ∈ (_ : Set _) := x.2
+  rw [Submodule.coe_sup] at mem1
+  rcases mem1 with ⟨a, b, a_mem, b_mem : b ∈ (Submodule.span R _ : Submodule R N), eq1⟩
+  rw [Submodule.mem_span_singleton] at b_mem
+  rcases b_mem with ⟨z, eq2⟩
+  exact ⟨⟨a, a_mem⟩, z, by rw [← eq1, ← eq2]⟩
```

Comment:
Is the idea that we don't need to #align private theorems?

Reply:
Fixed